gdal/gcore/gdal_misc.cpp

```diff
@@ -77,132 +77,163 @@ GDALDataType CPL_STDCALL
 GDALDataTypeUnion( GDALDataType eType1, GDALDataType eType2 )
 
 {
-    bool bSigned = false;
-    bool bFloating = false;
-    int nBits = 0;
+    const int panBits[] = { 
+        GDALGetDataTypeSizeBits( eType1 ), 
+        GDALGetDataTypeSizeBits( eType2 ) 
+    };
 
-    const bool bComplex = CPL_TO_BOOL(
-        GDALDataTypeIsComplex(eType1) | GDALDataTypeIsComplex(eType2) );
-
-    switch( eType1 )
-    {
-      case GDT_Byte:
-        nBits = 8;
-        bSigned = false;
-        bFloating = false;
-        break;
+    if(panBits[0] == 0 || panBits[1] == 0)
+        return GDT_Unknown;
+    
+    const bool pabSigned[] = { 
+        CPL_TO_BOOL( GDALDataTypeIsSigned( eType1 ) ),
+        CPL_TO_BOOL( GDALDataTypeIsSigned( eType2 ) )
+    };
+    
+    const bool bSigned = pabSigned[0] || pabSigned[1];
+    const bool bFloating =
+        CPL_TO_BOOL( GDALDataTypeIsFloating( eType1 ) ) ||
+        CPL_TO_BOOL( GDALDataTypeIsFloating( eType2 ) );
+    const bool bComplex = 
+        CPL_TO_BOOL( GDALDataTypeIsComplex( eType1 ) ) ||
+        CPL_TO_BOOL( GDALDataTypeIsComplex( eType2 ) );
+
+    const int nBits = pabSigned[0] != pabSigned[1] ?
+         std::max(panBits[pabSigned[1]], 2 * panBits[!pabSigned[1]]) :
```

Comment:
I have trouble understanding this logic. And this doesn't like to be symmetrical if you switch eType1 and eType2. Or at least this is not immediately obvious to my eyes